controllers/subnamespace_controller.go

```diff
@@ -34,18 +36,105 @@ type SubnamespaceReconciler struct {
+// +kubebuilder:rbac:groups=dana.dana.hns.io,resources=subnamespaces/status,verbs=get;update;patch
+
+func (r *SubnamespaceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("subspace", req.NamespacedName)
+	var subspace danav1alpha1.Subnamespace
+	var ownerNamespace v1.Namespace
+	var childNamespace v1.Namespace
+	var childNamespaceName string
+
+	// Getting reconciled subspace
+	if err := r.Get(ctx, req.NamespacedName, &subspace); err != nil {
+		log.Error(err, "Could not find Subspace")
+		return ctrl.Result{}, err
+	}
+	// Getting subspace's namespace - described as ownerNamespace
+	if err := r.Get(ctx, client.ObjectKey{Namespace: "", Name: subspace.GetNamespace()}, &ownerNamespace); err != nil {
+		log.Error(err, "Could not find owner Namespace ")
+		return ctrl.Result{}, err
+	}
+
+	// Hierarchy 'breadcrumb' name, built from ownerNamespace and subspace.
+	childNamespaceName = ownerNamespace.Name + "-" + subspace.Name
 
-func (r *SubnamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
-	_ = context.Background()
-	_ = r.Log.WithValues("subnamespace", req.NamespacedName)
+	// Subspace phase None flow
+	if subspace.Status.Phase == danav1alpha1.None {
 
-	// your logic here
+		if _, err := ctrl.CreateOrUpdate(ctx, r.Client, &subspace, func() error {
+			err := InitializeSubspace(ownerNamespace, subspace, childNamespaceName)
+			return err
+		}); err != nil {
+			log.Error(err, "Could not update Subspace Phase from 'None' to 'Missing'")
+		}
+	}
 
+	// Subspace phase Missing flow
+	if subspace.Status.Phase == danav1alpha1.Missing {
+		childNamespace = GetNewChildNamespace(ownerNamespace.Name, subspace.Name, childNamespaceName)
+		if err := r.Get(ctx, client.ObjectKey{Namespace: "", Name: childNamespaceName}, &childNamespace); err != nil {
+			if errors.IsNotFound(err) {
+				// Child Namespace is not found, therefore we are creating one
+				if err := r.Create(ctx, &childNamespace); err != nil {
+					log.Error(err, "Could not create Namespace ")
+					return ctrl.Result{}, err
+				}
+			}
+		} else {
+			// Child Namespace is exist, so we are changing the Subspace phase to created
+			if _, err := ctrl.CreateOrUpdate(ctx, r.Client, &subspace, func() error {
+				subspace.Status.Phase = danav1alpha1.Created
+				return nil
+			}); err != nil {
+				log.Error(err, "Could not update Subspace Phase from 'Missing' to 'Created'")
+			}
+		}
+
+	}
+
+	// Subspace phase Created flow
+	if subspace.Status.Phase == danav1alpha1.Created {
+		if err := r.Get(ctx, client.ObjectKey{Namespace: "", Name: childNamespaceName}, &childNamespace); err != nil {
+			if errors.IsNotFound(err) {
+				if _, err := ctrl.CreateOrUpdate(ctx, r.Client, &subspace, func() error {
+					subspace.Status.Phase = danav1alpha1.Missing
+					return nil
+				}); err != nil {
+					log.Error(err, "Could not update Subspace Phase from 'Created' to 'Missing'")
+				}
+			}
+		}
+	}
 	return ctrl.Result{}, nil
```

Comment:
when the ns controller deletes an SNS, I'm pretty sure it will go into this if statement is it ok or we need to address the delete case in the SNS controller?

Reply:
Do you think that if the NS controller will delete the SNS, it will be reconciled with the same values, with the same phase ?

Reply:
I think so, we should check before merging